### PR TITLE
Add nullbits TIDBIT

### PR DESCRIPTION
**Pull request by jaygreco**
## Description
Add the nullbits TIDBIT to mainline QMK. Up to this point I had maintained a [separate repo](https://github.com/nullbitsco/tidbit) with the keymaps for this board. Creating a PR for this now that our common file cleanup has been merged and I have some time. 

Most keymaps are unremarkable. The few exceptions are ~~the 14seg keymap, which includes files for interfacing with an I2C 14-segment alphanumeric display, and~~ the snailmap_lite keymap, which is a numpad port of [this awesome keymap](https://github.com/qmk/qmk_firmware/tree/master/keyboards/nullbitsco/nibble/keymaps/snailmap) for the NIBBLE. 

All keymaps have been tested extensively on my dev boards and are working with the latest QMK.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* N/A

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by jaygreco** [comment]
Checking in, I don't want this to go stale. Thanks!

File: keyboards/nullbitsco/tidbit/keymaps/14seg/HT16K33.h [hunk 1]
```diff
@@ -0,0 +1,94 @@
+/* Copyright 2021 Jay Greco
```
Comment (drashna) [inline keyboards/nullbitsco/tidbit/keymaps/14seg/HT16K33.h:1]: please rename this file to be lower cased.

File: keyboards/nullbitsco/tidbit/keymaps/14seg/keymap.c [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+                 ___, ___, ___, 
+  ___, ___, ___, ___, ___, ___, 
+  ___, ___, ___, ___, ___, ___, 
+  ___, ___, ___, ___, ___, ___, 
+  ___, ___, ___, ___, ___, MO(_NAV)   
+  ),
+};
+
+void matrix_init_user(void) {
+  matrix_init_remote_kb();
```
Comment (drashna) [inline keyboards/nullbitsco/tidbit/keymaps/14seg/keymap.c:80]: this should be called at the kb level, not here. 
```suggestion
```

File: keyboards/nullbitsco/tidbit/keymaps/14seg/keymap.c [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+  animation->message = message;
+  animation->mode = DISP_MODE_BOUNCE;
+  animation->enabled = true;
+
+  char tmp[] = "TIDBIT";
+  strcpy(message, tmp);
+}
+
+void matrix_scan_user(void) {
+  matrix_scan_remote_kb();
```
Comment (drashna) [inline keyboards/nullbitsco/tidbit/keymaps/14seg/keymap.c:95]:
```suggestion
```

File: keyboards/nullbitsco/tidbit/keymaps/14seg/keymap.c [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    if (timer_elapsed(tap_timer) > 120) {
+      //change layer
+      layer_on(_NAV);
+      nav_press_active = false;
+    }
+  }
+}
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  process_record_remote_kb(keycode, record);
```
Comment (drashna) [inline keyboards/nullbitsco/tidbit/keymaps/14seg/keymap.c:109]:
```suggestion

File: keyboards/nullbitsco/tidbit/keymaps/4rotary/config.h [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+// Encoders are defined in order. 1: B2 & B3, 2: B4 & B5, 3: D0 & D1, 4: D2 & D3
```
Comment (drashna) [inline keyboards/nullbitsco/tidbit/keymaps/4rotary/config.h:17]:
```suggestion
#pragma once

// Encoders are defined in order. 1: B2 & B3, 2: B4 & B5, 3: D0 & D1, 4: D2 & D3
```

File: keyboards/nullbitsco/tidbit/keymaps/4rotary/keymap.c [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+                 ___, ___, ___, 
+  ___, ___, ___, ___, ___, ___, 
+  ___, ___, ___, ___, ___, ___, 
+  ___, ___, ___, ___, ___, ___, 
+  ___, ___, ___, ___, ___, ___  
+  ),
+};
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  process_record_remote_kb(keycode, record);
```
Comment (drashna) [inline keyboards/nullbitsco/tidbit/keymaps/4rotary/keymap.c:45]:
```suggestion
```

File: keyboards/nullbitsco/tidbit/keymaps/4rotary/keymap.c [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+      register_code(KC_NLCK);
+    }
+    numlock_set = true;
+  }
+
+  return true;
+}
+
+void matrix_init_user(void) {
+  matrix_init_remote_kb();
```
Comment (drashna) [inline keyboards/nullbitsco/tidbit/keymaps/4rotary/keymap.c:62]:
```suggestion
```

File: keyboards/nullbitsco/tidbit/keymaps/4rotary/keymap.c [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+  set_bitc_LED(LED_OFF);
+}
+
+void matrix_scan_user(void) {
+  matrix_scan_remote_kb();
+}
```
Comment (drashna) [inline keyboards/nullbitsco/tidbit/keymaps/4rotary/keymap.c:68]:
```suggestion
```

File: keyboards/nullbitsco/tidbit/keymaps/default/keymap.c [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+                 ___, ___, ___, 
+  ___, ___, ___, ___, ___, ___, 
+  ___, ___, ___, ___, ___, ___, 
+  ___, ___, ___, ___, ___, ___, 
+  ___, ___, ___, ___, ___, ___  
+  ),
+};
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  process_record_remote_kb(keycode, record);
```
Comment (drashna) [inline keyboards/nullbitsco/tidbit/keymaps/default/keymap.c:45]:
```suggestion
```

File: keyboards/nullbitsco/tidbit/keymaps/default/keymap.c [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+      register_code(KC_NLCK);
+    }
+    numlock_set = true;
+  }
+
+  return true;
+}
+
+void matrix_init_user(void) {
+  matrix_init_remote_kb();
```
Comment (drashna) [inline keyboards/nullbitsco/tidbit/keymaps/default/keymap.c:62]:
```suggestion
```

File: keyboards/nullbitsco/tidbit/keymaps/default/keymap.c [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+  set_bitc_LED(LED_OFF);
+}
+
+void matrix_scan_user(void) {
+  matrix_scan_remote_kb();
+}
```
Comment (drashna) [inline keyboards/nullbitsco/tidbit/keymaps/default/keymap.c:68]:
```suggestion
```

File: keyboards/nullbitsco/tidbit/keymaps/oled/keymap.c [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+                 ___, ___, ___, 
+  ___, ___, ___, ___, ___, ___, 
+  ___, ___, ___, ___, ___, ___, 
+  ___, ___, ___, ___, ___, ___, 
+  ___, ___, ___, ___, ___, ___
+  ),
+};
+
+
+#ifdef OLED_ENABLE
```
Comment (drashna) [inline keyboards/nullbitsco/tidbit/keymaps/oled/keymap.c:67]: please move the oled code to the keyboard c file, and use the oled_task_kb instead

File: keyboards/nullbitsco/tidbit/keymaps/oled/keymap.c [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+void matrix_init_user(void) { 
+    matrix_init_remote_kb();
+}
+
+void matrix_scan_user(void) {
+    matrix_scan_remote_kb();
+}
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+    process_record_remote_kb(keycode, record);
```
Comment (drashna) [inline keyboards/nullbitsco/tidbit/keymaps/oled/keymap.c:122]:
```suggestion
bool process_record_user(uint16_t keycode, keyrecord_t *record) {
```

File: keyboards/nullbitsco/tidbit/keymaps/oled/keymap.c [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+    if (!numlock_set && record->event.pressed) {
+        led_t led_state = host_keyboard_led_state();
+        if (!led_state.num_lock) {
+            register_code(KC_NLCK);
+        }
+        numlock_set = true;
+    }
+
+    switch (keycode) {
+        case PROG:
```
Comment (drashna) [inline keyboards/nullbitsco/tidbit/keymaps/oled/keymap.c:136]:
```suggestion
        case RESET:
```

File: keyboards/nullbitsco/tidbit/keymaps/oled/keymap.c [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+            if (record->event.pressed) {
+                set_bitc_LED(LED_DIM);
+                rgblight_disable_noeeprom();
+                oled_off();
+                bootloader_jump();  // jump to bootloader
+            }
+            break;
```
Comment (drashna) [inline keyboards/nullbitsco/tidbit/keymaps/oled/keymap.c:143]:
```suggestion
            return false;
```

File: keyboards/nullbitsco/tidbit/keymaps/14seg/config.h [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+// Encoders are defined in order. 1: B2 & B3, 2: B4 & B5, 3: D0 & D1, 4: D2 & D3
+#define ENCODERS_PAD_A { B2, B4 }
```
Comment (drashna) [inline keyboards/nullbitsco/tidbit/keymaps/14seg/config.h:19]: these should be defined at the keyboard level

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by jaygreco** [comment]
I added two commits: the first removes keymaps that weren't worth including after thinking about it. The second implements the PR feedback. I also cleaned up the file formatting and indentation to adhere to the QMK coding convention.

File: keyboards/nullbitsco/tidbit/keymaps/oled/keymap.c (hunk 12 above)
Comment (jaygreco) [reply keyboards/nullbitsco/tidbit/keymaps/oled/keymap.c:63]: This was the only change I wasn't sure about. There is more than one keymap with OLED functionality -- would moving the OLED code to the keyboard file mean it applies to all keymaps with the OLED enabled?

**Comment by jaygreco** [comment]
Thanks @drashna for the PR feedback!

File: keyboards/nullbitsco/tidbit/keymaps/oled/keymap.c (hunk 12 above)
Comment (drashna) [reply keyboards/nullbitsco/tidbit/keymaps/oled/keymap.c:63]: Moving the code to the keyboard's c file allows any of the keymaps to have that set as the default, but can allow for the display to be overridden by users.   It also allows the board to have a default config for the QMK Configurator.

Comment (jaygreco) [reply keyboards/nullbitsco/tidbit/keymaps/oled/keymap.c]: Ah, that makes sense. Changed in the latest revision. Thanks!

**Comment by jaygreco** [comment]
@drashna great advice on moving a lot of this stuff to the keyboard file. I pulled out more beyond what you'd recommended in the latest commit. Thanks for the feedback!

File: keyboards/nullbitsco/tidbit/keymaps/snailmap_lite/keymap.c [hunk 17]
```diff
@@ -0,0 +1,294 @@
...
+    uint8_t cursorposition_cur = 13;
+    uint8_t cursorposition_max = 14;
+
+    char wpm_counter[4];
+    wpm_counter[3] = '\0';
+    wpm_counter[2] = '0' + current_wpm % 10;
+    wpm_counter[1] = '0' + (current_wpm / 10) % 10;
+    wpm_counter[0] = '0' + (current_wpm / 100) % 10;
+    oled_set_cursor(0, cursorposition_cur);
+    oled_write(wpm_counter, false);
```
Comment (drashna) [inline keyboards/nullbitsco/tidbit/keymaps/snailmap_lite/keymap.c:169]:
```suggestion
    oled_set_cursor(0, cursorposition_cur);
    oled_write(get_u8_str(current_wpm), false);
```

File: keyboards/nullbitsco/tidbit/rules.mk [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+MIDI_ENABLE = no            # MIDI support
+AUDIO_ENABLE = no           # Audio output on port C6
+ENCODER_ENABLE = yes        # Use rotary encoder
+LTO_ENABLE = yes            # Link-time optimization
+
+# Project specific files
+SRC += i2c_master.c \
+    common/bitc_led.c \
+    common/remote_kb.c \
+    uart.c
```
Comment (drashna) [inline keyboards/nullbitsco/tidbit/rules.mk:28]:
```suggestion
SRC += common/bitc_led.c \
    common/remote_kb.c
QUANTUM_LIB_SRC += i2c_master.c \
    uart.c
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by jaygreco** [comment]
@drashna Feedback incorporated. I also removed the WPM fudge factor from the same keymap at the same time.

**Review by drashna — APPROVED** [review]

File: keyboards/nullbitsco/tidbit/rules.mk [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
+MIDI_ENABLE = no            # MIDI support
+AUDIO_ENABLE = no           # Audio output on port C6
+ENCODER_ENABLE = yes        # Use rotary encoder
+LTO_ENABLE = yes            # Link-time optimization
```
Comment (fauxpark) [inline keyboards/nullbitsco/tidbit/rules.mk:22]:
```suggestion
BOOTMAGIC_ENABLE = no       # Enable Bootmagic Lite
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
NKRO_ENABLE = no            # Enable N-Key Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
AUDIO_ENABLE = no           # Audio output
ENCODER_ENABLE = yes
LTO_ENABLE = yes
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by jaygreco** [comment]
Hi @drashna and @fauxpark, apologies for the ping! I'm just checking in on this. I'm hoping to get it merged before the breaking changes push this month. Thanks!

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**